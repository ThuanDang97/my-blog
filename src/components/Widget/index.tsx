import { memo } from 'react'
import {
  Box,
  BoxProps,
  List,
  ListItem,
  Typography,
  useTheme,
} from '@mui/material'
import Link from 'next/link'

export interface IWidgetsItem {
  name: string
  slug: string
}

export interface IWidget extends BoxProps {
  title: string
  widgetsItem: IWidgetsItem[]
}
const Widget = ({ title = 'Title', widgetsItem, ...props }: IWidget) => {
  const theme = useTheme()
  return (
    <Box
      component="aside"
      sx={{
        border: '1px solid #EAEBED',
        padding: '25px',
        marginBottom: '50px',
        position: 'relative',
        clear: 'both',
        float: 'left',
        width: '100%',
      }}
      {...props}
    >
      <Typography
        data-testid="widget-title"
        variant="h2"
        sx={{
          fontSize: '25px',
          position: 'absolute',
          textTransform: 'capitalize',
          fontWeight: 600,
          top: '-17px',
          background: '#fff',
          marginLeft: '-26px',
          display: 'inline-block',
          paddingRight: '30px',
        }}
      >
        {title}
      </Typography>
      <List>
        {widgetsItem.map((item) => (
          <ListItem
            data-testid="widget-item"
            key={item.slug}
            sx={{
              fontSize: '16px',
              fontWeight: 'bold',
              ':hover': {
                color: theme.palette.grey[500],
              },
            }}
          >
            <Link href={`/${item.slug}`}>{item.name}</Link>
          </ListItem>
        ))}
      </List>
    </Box>
  )
}

export default memo(Widget)
