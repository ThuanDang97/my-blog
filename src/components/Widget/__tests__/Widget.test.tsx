// Utils
import { cleanup, render, screen } from '@utils/testUtils'

// Mocks data
import { MOCK_WIDGET } from '@constants/__mock__/mockData'

// Components
import Widget, { IWidget } from '..'

describe('Widget component', () => {
  afterEach(cleanup)

  const mockProps: IWidget = MOCK_WIDGET

  it('should render Widget component', () => {
    const { container } = render(<Widget {...mockProps} />)

    expect(container).toMatchSnapshot()

    const title = screen.getByTestId('widget-title')
    expect(title).toHaveTextContent('Categories')
  })

  it('should render correct widget item', () => {
    render(<Widget {...mockProps} />)

    const link = screen.getAllByTestId('widget-item')
    const links = screen.getAllByRole('link')

    expect(link[0]).toHaveTextContent('Python')
    expect(links[0].getAttribute('href')).toEqual('/python')
  })
})
