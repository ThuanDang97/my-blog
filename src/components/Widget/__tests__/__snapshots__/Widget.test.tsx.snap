// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widget component should render Widget component 1`] = `
<div>
  <aside
    class="MuiBox-root css-9lsm0f"
  >
    <h2
      class="MuiTypography-root MuiTypography-h2 css-w1deiu-MuiTypography-root"
      data-testid="widget-title"
    >
      Categories
    </h2>
    <ul
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1otctin-MuiListItem-root"
        data-testid="widget-item"
      >
        <a
          href="/python"
        >
          Python
        </a>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1otctin-MuiListItem-root"
        data-testid="widget-item"
      >
        <a
          href="/javascript"
        >
          JavaScript
        </a>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1otctin-MuiListItem-root"
        data-testid="widget-item"
      >
        <a
          href="/nextjs"
        >
          NextJS
        </a>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1otctin-MuiListItem-root"
        data-testid="widget-item"
      >
        <a
          href="/front-end"
        >
          Front-End
        </a>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1otctin-MuiListItem-root"
        data-testid="widget-item"
      >
        <a
          href="/linux"
        >
          Linux
        </a>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1otctin-MuiListItem-root"
        data-testid="widget-item"
      >
        <a
          href="/css"
        >
          CSS
        </a>
      </li>
    </ul>
  </aside>
</div>
`;
